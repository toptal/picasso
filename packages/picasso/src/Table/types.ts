export type TableSpacing = 'compact' | 'narrow' | 'regular'
export type TableVariant = 'clear' | 'bordered' | 'striped'
