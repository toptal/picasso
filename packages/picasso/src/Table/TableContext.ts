import { createContext } from 'react'

export interface TableConfig {
  spacing: 'compact' | 'narrow' | 'regular'
  variant: 'clear' | 'bordered' | 'striped'
}

const TableContext = createContext<TableConfig>({
  spacing: 'regular',
  variant: 'bordered'
})

export default TableContext
