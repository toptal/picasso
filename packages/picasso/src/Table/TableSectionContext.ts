import { createContext } from 'react'

export enum TableSection {
  HEAD,
  BODY,
  FOOTER
}

export const TableSectionContext = createContext<TableSection | undefined>(
  undefined
)
