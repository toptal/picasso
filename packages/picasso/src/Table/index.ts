import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Table'
export { TableSectionContext, TableSection } from './TableSectionContext'
export { default as TableContext } from './TableContext'
export type { TableConfig } from './TableContext'

export { default } from './Table'
export type TableProps = OmitInternalProps<Props>
