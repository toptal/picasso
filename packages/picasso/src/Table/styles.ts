import { createStyles } from '@material-ui/core/styles'
import { PicassoProvider } from '@toptal/picasso-shared'

PicassoProvider.override(() => ({}))

export default () => createStyles({})
