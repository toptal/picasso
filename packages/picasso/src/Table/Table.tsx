import React, {
  forwardRef,
  ReactNode,
  TableHTMLAttributes,
  useMemo,
} from 'react'
import { makeStyles, Theme } from '@material-ui/core/styles'
import MUITable from '@material-ui/core/Table'
import { BaseProps } from '@toptal/picasso-shared'

import TableContext from './TableContext'
import styles from './styles'

export interface Props
  extends BaseProps,
    TableHTMLAttributes<HTMLTableElement> {
  /** Children components (`Table.Head`, `Table.Body`, `Table.Footer`) */
  children: ReactNode
  /** Inner spacing */
  spacing?: 'compact' | 'narrow' | 'regular'
  /** Appearance variant */
  variant?: 'clear' | 'bordered' | 'striped'
}

const useStyles = makeStyles<Theme>(styles, {
  name: 'Table',
})

// eslint-disable-next-line react/display-name
export const Table = forwardRef<HTMLTableElement, Props>(function Table(
  props,
  ref
) {
  const {
    className,
    style,
    children,
    spacing = 'regular',
    variant = 'bordered',
    ...rest
  } = props
  const classes = useStyles()

  const tableConfig = useMemo(() => ({ spacing, variant }), [spacing, variant])

  return (
    <TableContext.Provider value={tableConfig}>
      <MUITable
        {...rest}
        ref={ref}
        classes={classes}
        className={className}
        style={style}
      >
        {children}
      </MUITable>
    </TableContext.Provider>
  )
})

Table.defaultProps = {
  spacing: 'regular',
  variant: 'bordered',
}

Table.displayName = 'Table'

export default Table
