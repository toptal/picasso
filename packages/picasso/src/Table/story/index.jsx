import tableCellStory from '../../TableCell/story'
import tableBodyStory from '../../TableBody/story'
import tableFooterStory from '../../TableFooter/story'
import tableHeadStory from '../../TableHead/story'
import tableSectionHeadStory from '../../TableSectionHead/story'
import tableRowStory from '../../TableRow/story'
import tableExpandableRowStory from '../../TableExpandableRow/story'
import { Table } from '../Table'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Components').createPage(
  'Table',
  'Display sets of data'
)

page
  .createTabChapter('Props')
  .addComponentDocs({
    component: Table,
    name: 'Table',
    description: 'Root component representing table',
    additionalDocs: {
      spacing: {
        name: 'spacing',
        type: { name: 'enum', enums: ['"regular"', '"compact"', '"narrow"'] },
        description: 'Inner spacing',
      },
      variant: {
        name: 'variant',
        type: { name: 'enum', enums: ['"clear"', '"bordered"', '"striped"'] },
        description: 'Appearance variant',
      },
    },
  })
  .addComponentDocs(tableHeadStory.componentDocs)
  .addComponentDocs(tableSectionHeadStory.componentDocs)
  .addComponentDocs(tableFooterStory.componentDocs)
  .addComponentDocs(tableBodyStory.componentDocs)
  .addComponentDocs(tableRowStory.componentDocs)
  .addComponentDocs(tableCellStory.componentDocs)
  .addComponentDocs(tableExpandableRowStory.componentDocs)

page
  .createChapter()
  .addExample('Table/story/Default.example.tsx', {
    title: 'Plain table',
    takeScreenshot: false,
  })
  .addExample('Table/story/Variants.example.tsx', {
    title: 'Appearance variants',
    takeScreenshot: false,
  })
  .addExample('Table/story/Spacings.example.tsx', {
    title: 'Inner spacing',
    takeScreenshot: false,
  })
  .addExample('Table/story/Alignments.example.tsx', {
    title: 'Cell alignments',
    takeScreenshot: false,
  })
  .addExample('Table/story/Select.example.tsx', {
    title: 'Selectable table',
    takeScreenshot: false,
  })
  .addExample(
    'Table/story/SectionHeader.example.tsx',
    'Table with section header'
  )
  .addExample(
    'Table/story/MultipleSectionHeader.example.tsx',
    'Table with multiple sections'
  )
  .addExample('Table/story/ExpandableRows.example.tsx', {
    title: 'Expandable rows',
    takeScreenshot: false,
  })
  .addExample('Table/story/ExpandableRowsDefaultExpanded.example.tsx', {
    title: 'Expandable rows, expanded by default',
    takeScreenshot: false,
  })

page.connect(tableCellStory.chapter)
