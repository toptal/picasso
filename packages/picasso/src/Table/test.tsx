import React from 'react'
import { render, PicassoConfig } from '@toptal/picasso/test-utils'
import * as titleCaseModule from 'ap-style-title-case'

import { TableProps } from '../Table'
import { TableCompound as Table } from '../TableCompound'
import TableContext from './TableContext'
import { TableCellProps } from '../TableCell'

jest.mock('ap-style-title-case')

const DEFAULT_CELLS_PROPS: [TableCellProps, TableCellProps, TableCellProps] = [
  { children: 'Table test' },
  { children: 'Table test' },
  { children: 'Table test' },
]

const renderTable = (
  picassoConfig: PicassoConfig = { titleCase: false },
  cellsProps: [
    TableCellProps,
    TableCellProps,
    TableCellProps
  ] = DEFAULT_CELLS_PROPS,
  tableProps: Pick<
    TableProps,
    'className' | 'style' | 'spacing' | 'variant'
  > = {}
) => {
  return render(
    <Table {...tableProps}>
      <Table.Head>
        <Table.Row>
          <Table.Cell titleCase={cellsProps[0].titleCase}>
            {cellsProps[0].children}
          </Table.Cell>
        </Table.Row>
      </Table.Head>
      <Table.Body>
        <Table.Row>
          <Table.Cell titleCase={cellsProps[1].titleCase}>
            {cellsProps[1].children}
          </Table.Cell>
        </Table.Row>
      </Table.Body>
      <Table.Footer>
        <Table.Row>
          <Table.Cell titleCase={cellsProps[2].titleCase}>
            {cellsProps[2].children}
          </Table.Cell>
        </Table.Row>
      </Table.Footer>
    </Table>,
    undefined,
    picassoConfig
  )
}

const spiedOnTitleCase = jest.spyOn(titleCaseModule, 'default')

describe('Table', () => {
  beforeEach(() => {
    spiedOnTitleCase.mockReset()
  })

  afterEach(() => {
    spiedOnTitleCase.mockReset()
  })

  it('renders', () => {
    const { container } = renderTable()

    expect(container).toMatchSnapshot()
  })

  it('sets attributes correctly', () => {
    const { getByTestId } = render(
      <Table
        data-testid='table'
        className='foo'
        spacing='compact'
        variant='bordered'
      >
        <TableContext.Consumer>
          {({ variant, spacing }) => (
            <Table.Body>
              <Table.Row>
                <Table.Cell data-testid='variant'>{variant}</Table.Cell>
                <Table.Cell data-testid='spacing'>{spacing}</Table.Cell>
              </Table.Row>
            </Table.Body>
          )}
        </TableContext.Consumer>
      </Table>
    )

    const table = getByTestId('table')

    expect(getByTestId('spacing').textContent).toBe('compact')
    expect(getByTestId('variant').textContent).toBe('bordered')
    expect(table?.classList.contains('foo')).toBeTruthy()
  })

  describe('when Picasso titleCase property is true', () => {
    describe('when component does not override titleCase property', () => {
      it('should transform table header text to title case', () => {
        renderTable({ titleCase: true })

        expect(spiedOnTitleCase).toHaveBeenCalledTimes(1)
        expect(spiedOnTitleCase).toHaveBeenCalledWith(
          DEFAULT_CELLS_PROPS[0].children
        )
      })
    })

    describe('when component overrides titleCase property', () => {
      it('should not transform table header text to title case', () => {
        const cellsProps = DEFAULT_CELLS_PROPS.map(cellProps => ({
          ...cellProps,
          titleCase: false,
        })) as typeof DEFAULT_CELLS_PROPS

        renderTable({ titleCase: true }, cellsProps)

        expect(spiedOnTitleCase).toHaveBeenCalledTimes(0)
      })
    })
  })
})
