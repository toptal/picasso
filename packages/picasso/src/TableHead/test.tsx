import React from 'react'
import { render, cleanup, RenderResult } from '@testing-library/react'
import Picasso from '@toptal/picasso-shared'

import Table from '../Table'

const renderTableHead = () => {
  return render(
    <Picasso loadFonts={false}>
      <Table>
        <Table.Head>
          <Table.Row>
            <Table.Cell>Head test</Table.Cell>
          </Table.Row>
        </Table.Head>
      </Table>
    </Picasso>
  )
}

afterEach(cleanup)

describe('TableHead', () => {
  let api: RenderResult

  beforeEach(() => {
    api = renderTableHead()
  })

  test('default render', () => {
    const { container } = api

    expect(container).toMatchSnapshot()
  })
})
