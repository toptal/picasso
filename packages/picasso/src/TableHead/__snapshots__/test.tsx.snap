// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHead renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <table
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root TableRow-root MuiTableRow-head TableRow-head"
        >
          <th
            class="MuiTableCell-root TableCell-root MuiTableCell-head TableCell-head"
            scope="col"
          >
            Head test
          </th>
        </tr>
      </thead>
    </table>
  </div>
</div>
`;
